core/src/main/java/org/apache/iceberg/RollingManifestWriter.java

```diff
@@ -0,0 +1,232 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.util.List;
+import org.apache.iceberg.io.FileAppender;
+import org.apache.iceberg.io.FileIO;
+import org.apache.iceberg.io.OutputFile;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+
+/** As opposed to {@link ManifestWriter}, a rolling writer could produce multiple manifest file. */
+abstract class RollingManifestWriter<T extends ContentFile<T>> implements FileAppender<T> {
```

Comment:
@szehon-ho pls take a look if this is the right direction. Here implements `FileAppender` instead of `FileWriter` that's because the `ManifestWriter` implements `FileAppender`.

Reply:
Thanks, this definitely looks cleaner. 

Im just not sure what to do with existing hierarchy (ManifestWriter), as there are many duplications.  

Did we think about modifying ManifestWriter, to take extra variable 'targetFileSizeBytes', then trigger tryRollover only in this case.

The problem may be the method manifestFile() in writer?  It should have been manifestFiles().  Maybe we can deprecate existing method?

CC @aokolnychyi .

Reply:
Actually on second read, this may be ok, as its a composition, and not duplicating that much code.  Lets see if Anton has any comments